// sol/lib/sol.js
/*!
 * Sol
 * Copyright(c) 2011-2012 Jake Luer <jake@alogicalparadox.com>
 * MIT Licensed
 */

function merge (a, b) {
  if (a && b) {
    for (var key in b) {
      a[key] = b[key];
    }
  }
  return a;
};

/*!
 * Basic comparators
 */

var comparator = {};

comparator.ASC = function (a, b) {
  return a.value - b.value;
};

comparator.DESC = function (a, b) {
  return b.value - a.value;
};

comparator.KASC = function (a, b) {
  var A = a.key.toLowerCase()
    , B = b.key.toLowerCase();
  if (A < B) return -1;
  else if (A > B) return 1;
  else return 0;
};

comparator.KDESC= function (a, b) {
  var A = a.key.toLowerCase()
    , B = b.key.toLowerCase();
  if (A < B) return 1;
  else if (A > B) return -1;
  else return 0;
};

/*!
 * Main export
 */

module.exports = Sol;

/**
 * # Sol (constructor)
 *
 * Creates a new hash given a set
 * of values and options.
 *
 * @param {Array} values to insert
 * @param {Object} options
 * @api public
 * @name constructor
 */

function Sol (values, opts) {
  this.opts = opts || {};
  this._data = Object.create(null);
  for (var key in values) {
    this.set(key, values[key]);
  }
}

/*!
 * Version
 */

Sol.version = '0.3.0';

/**
 * # length
 *
 * Returns the the number of objects in the hash
 *
 * @returns {Number}
 * @api public
 * @name length
 */

Object.defineProperty(Sol.prototype, 'length',
  { get: function () {
      return this.keys.length;
    }
});

/**
 * # keys
 *
 * Returns an Array of all keys in the hash.
 *
 * @returns {Array} keys
 * @api public
 * @name keys
 */

Object.defineProperty(Sol.prototype, 'keys',
  { get: function () {
      var keys = []
        , value;
      for (var key in this._data) {
        value = this._data[key];
        if ('undefined' != typeof value)
          keys.push(key);
      }
      return keys;
    }
});

/**
 * # values
 *
 * Returns an array of all values in the hash.
 *
 * @returns {Array} values
 * @api public
 * @name values
 */

Object.defineProperty(Sol.prototype, 'values',
  { get: function () {
      var vals = []
        , value;
      for (var key in this._data) {
        value = this._data[key];
        if ('undefined' != typeof value)
          vals.push(value);
      }
      return vals;
    }
});

/**
 * # .get(key)
 *
 * Gets the value of at a specific key.
 *
 * @param {String|Number} key
 * @returns {Object} value
 * @api public
 * @name .get()
 */

Sol.prototype.get = function (key) {
  return this._data[key];
};

/**
 * # .has(key)
 *
 * Checks for existence of key in hash.
 *
 * @param {String} key
 * @returns {Boolean} existence
 * @api public
 * @name .has()
 */

Sol.prototype.has = function (key) {
  var val = this._data[key];
  return !!val;
};

/**
 * # .set(key, value)
 *
 * Sets the value at a given key.
 *
 * @param {String} key
 * @param {Object} value
 * @api public
 * @name .set()
 */

Sol.prototype.set = function (key, value) {
  this._data[key] = value;
};

/**
 * # .del(key, [silent])
 *
 * Removes a key from the Sol.
 *
 * @param {String} key
 * @api public
 * @name .del()
 */

Sol.prototype.del = function (key) {
  this._data[key] = undefined;
};

/**
 * # .clone()
 *
 * Returns a copy of the hash.
 *
 * @returns {Sol} cloned
 * @api public
 * @name .clone()
 */

Sol.prototype.clone = function () {
  var hash = new this.constructor(null, this.opts);
  merge(hash._data, this._data);
  return hash;
};

/**
 * # .at(index)
 *
 * Returns the data at a given index,
 * as if the hash as an array.
 *
 * @param {Number} index
 * @returns {Object} value at index
 * @api public
 * @name .at()
 */

Sol.prototype.at = function (index) {
  var key = this.keys[index];
  return this._data[key];
};

/**
 * # .index(key)
 *
 * Returns the index for a given
 * key, as if the hash was an array.
 *
 * @param {String} key
 * @returns {Number} index
 * @api public
 * @name .index()
 */

Sol.prototype.index = function (key) {
  return this.keys.indexOf(key);
};

/**
 * # .each(iterator, [context])
 *
 * Apply a given iteration function to
 * each value in the hash.
 *
 * @param {Function} iterator
 * @param {Object} context to apply as `this` to iterator. Defaults to current hash.
 * @alias forEach
 * @api public
 * @name .each()
 */

Sol.prototype.each =
Sol.prototype.foEach = function (iterator, context) {
  var index = 0;
  context = context || this;
  for (var key in this._data) {
    iterator.call(context, this._data[key], key, index);
    index++;
  }
};

/**
 * # .map(iterator, [context])
 *
 * Map a given function to every value
 * in the Sol. Iterator must return new value.
 *
 * @param {Function} iterator
 * @param {Object} content to apply as `this` to iterator. Defults to current hash.
 * @api public
 * @name .map()
 */

Sol.prototype.map = function (iterator, context) {
  var hash = merge({}, this._data)
    , context = context || this;

  this.each(function(value, key, index) {
    hash[key] = iterator.call(context, hash[key], key, index);
  });

  return new this.constructor(hash);
};

/**
 * # .reduce(iterator, [context])
 *
 * @param {Function} iterator
 * @param {Object} context to apply as `this` to the iterator. Defaults to current hash.
 * @returns {Number} result
 * @api public
 * @name .reduce()
 */

Sol.prototype.reduce = function (iterator, initialValue, context) {
  if ('number' !== typeof initialValue) {
    context = initialValue;
    initialValue = 0;
  }

  var res = initialValue || 0
    , context = context || this;

  this.each(function (value, key, index) {
    res += iterator.call(context, value, key, index);
  });

  return res;
};

/**
 * # .mapReduce(mapFn, reduceFn)
 *
 * @param {Function} map function
 * @param {Function} reduce function
 * @returns Sol
 * @api public
 * @name .mapReduce()
 */

Sol.prototype.mapReduce = function (mapFn, reduceFn) {
  var hash = merge({}, this._data)
    , mapSol = new this.constructor(null, this.opts)
    , reduceSol = new this.constructor(null, this.opts);

  this.each(function (_value, _key) {
    mapFn(_key, _value, function emit (key, value) {
      if (mapSol.has(key)) mapSol.get(key).push(value);
      else mapSol.set(key, [ value ]);
    });
  });

  mapSol.each(function (_value, key) {
    var value = reduceFn(key, _value);
    reduceSol.set(key, value);
  });

  return reduceSol;
};

/**
 * # .filter(iterator, [context])
 *
 * Creates a new hash with all key/values
 * that pass a given iterator test. Iterator
 * should return `true` or `false`.
 *
 * @param {Function} iterator
 * @param {Object} context to apply as `this` to iterator. Defaults to current hash.
 * @api public
 * @name .filter()
 */

Sol.prototype.filter = function (iterator, context) {
  var hash = new this.constructor(null, this.opts)
    , context = context || this;
  this.each(function(value, key, index) {
    if (iterator.call(context, value, key, index))
      hash.set(key, value);
  });
  return hash;
};

/**
 * # .sort(iterator)
 *
 * Creates a new hash sorted with an iterator
 *
 * Options:
 * - **kasc**: key alphabetical (default)
 * - **kdesc**: key alphabetical descending
 * - **asc**: value ascending, only if value is not object
 * - **desc**: value descending, only if value is not object
 * - **{Function}**: provide custom iterator
 *
 * Custom Iterator:
 *
 *    hash.sort(function (a, b) {
 *      var akey = a.key
 *        , avalue = a.value;
 *      // etc..
 *      return 1 || -1 || 0; // choose
 *    });
 *
 * @param {String|Function} iterator
 * @returns {this} for chaining
 * @api public
 * @name .sort()
 */

Sol.prototype.sort = function (iter) {
  var iterator;

  // default to key asc
  if (!iter) iter = 'kasc';

  // if we are passed a custom iterator
  if ('function' == typeof iter) iterator = iter;

  // else we have a string and an iterator
  else if ('string' == typeof iter) {
    iter = iter.toUpperCase();
    if (comparator[iter]) iterator = comparator[iter];
  }

  var arr = this.toArray().sort(iterator)
  this.flush(true);
  this.fromArray(arr);
  return this;
};

/**
 * # .clean()
 *
 * Helper function to remove all keys that have an
 * `undefined` value. Useful, as `del` leaves keys.
 *
 * @api public
 * @name clean
 */

Sol.prototype.clean = function () {
  var arr = this.toArray();
  this.flush(true);
  this.fromArray(arr);
  return this;
};

/**
 * # .flush(silent)
 *
 * Helper function to remove all values from a hash.
 *
 * @api public
 * @name flush
 */

Sol.prototype.flush = function () {
  delete this._data;
  this._data = Object.create(null);
  return this;
};

/**
 * # .toArray()
 *
 * Returns the hash as javascript array in
 * the format `{ key: [key], value: [value] }`.
 *
 * @returns {Array}
 * @api public
 * @name .toArray()
 */

Sol.prototype.toArray = function () {
  var self = this
    , arr = [];

  this.each(function (value, key) {
    if (self.has(key)) {
      var obj = { key: key, value: value };
      arr.push(obj);
    }
  });

  return arr;
};

/**
 * # .fromArray()
 *
 * Sets values in this Hash from the result of
 * of a `sol.toArray` call. Array must be in
 * the format `{ key: [key], value: [value] }`.
 *
 * @returns {this}
 * @api public
 * @name fromArray
 */

Sol.prototype.fromArray = function (arr) {
  for (var i = 0; i < arr.length; i++) {
    var key = arr[i].key
      , val = arr[i].value;
    if ('undefined' !== typeof val) {
      this.set(key, val);
    }
  }

  return this;
};
